tool/tctl/common/idp_command.go

```diff
@@ -0,0 +1,273 @@
+		&samlIdPCommand{},
+	}
+
+	for _, subcommandRunner := range t.subcommandRunners {
+		subcommandRunner.initialize(idp, cfg)
+	}
+}
+
+// TryRun calls tryRun for each subcommand, and returns (false, nil) if none of them match.
+func (i *IdPCommand) TryRun(ctx context.Context, cmd string, c auth.ClientI) (match bool, err error) {
+	for _, subcommandRunner := range i.subcommandRunners {
+		match, err = subcommandRunner.tryRun(ctx, cmd, c)
+		if err != nil {
+			return match, trace.Wrap(err)
+		}
+		if match {
+			return match, nil
+		}
+	}
+	return false, nil
+}
+
+func (s *samlIdPCommand) initialize(parent *kingpin.CmdClause, cfg *servicecfg.Config) {
+	samlcmd := parent.Command("saml", "SAML Identity Provider")
+	samlcmd.Alias(`
+Examples:
+
+  Test attribute mapping with given user and service provider.
+  > tctl idp saml test_attribute_mapping --users user.yaml (user spec file or username) --sp sp.yaml (service provider spec file) --format (json,yaml)
+`)
+	s.cmd = samlcmd
+
+	testAttrMap := samlcmd.Command("test_attribute_mapping", "Test expression evaluation of attribute mapping.")
+	testAttrMap.Flag("users", "username or name of a file containing user spec").Short('u').StringsVar(&s.testAttributeMapping.users)
+	testAttrMap.Flag("sp", "name of a file containing service provider spec").StringVar(&s.testAttributeMapping.serviceProvider)
+	testAttrMap.Flag("format", "output format, 'yaml' or 'json'").StringVar(&s.testAttributeMapping.outFormat)
+	testAttrMap.Alias(`
+Examples:
+
+	# test with username and service provider file
+	> tctl idp saml test_attribute_mapping --users user1 --sp sp.yaml
+
+	# test with multiple usernames and service provider file
+	> tctl idp saml test_attribute_mapping --users user1,user2 --sp sp.yaml
+
+	# test with user and service provider file and print output in yaml format
+	> tctl idp saml test_attribute_mapping --users user1.yaml --sp sp.yaml --format yaml
+`)
+	s.testAttributeMapping.cmd = testAttrMap
+}
+
+func (s *samlIdPCommand) tryRun(ctx context.Context, cmd string, c auth.ClientI) (match bool, err error) {
+	switch cmd {
+	case s.testAttributeMapping.cmd.FullCommand():
+		return true, trace.Wrap(s.testAttributeMapping.run(ctx, c))
+	default:
+		return false, nil
+	}
+}
+
+// testCommand implements the "tctl idp saml test_attribute_mapping" command.
+type testAttributeMapping struct {
+	cmd *kingpin.CmdClause
+
+	serviceProvider string
+	users           []string
+	outFormat       string
+}
+
+func (t *testAttributeMapping) run(ctx context.Context, c auth.ClientI) error {
+	if len(t.serviceProvider) == 0 && len(t.users) == 0 {
+		return trace.BadParameter("no attributes to test, --users and --sp must be set")
+	}
+	if len(t.users) == 0 {
+		return trace.BadParameter("--users must be set. Either provide username or name of a file containing user spec")
+	}
+	if len(t.serviceProvider) == 0 {
+		return trace.BadParameter("--sp must be set. Provide name of a file containing service provider spec")
+	}
```

Comment:
Can we just use build-in cli lib `Reuired()` mechanism: 

```go
testAttrMap.Flag("users", "...").Required().Short('u').StringsVar(&s.testAttributeMapping.users)
testAttrMap.Flag("sp", "...").Required.().StringVar(&s.testAttributeMapping.serviceProvider)
```

Reply:
Oh thanks, didn't dig it enough to check `Required()` is available, used in https://github.com/gravitational/teleport/pull/36360/commits/560dd564ed1cc102e1ff7922cbdd67f0bdcd9797